CockroachDB support

chappjc opened the pull request:
<s>Resolves</s> Working toward https://github.com/decred/dcrdata/issues/1291, the following changes are made to support CockroachDB (crdb):

- Detecting use of cockroach.
- Removal of custom PostgreSQL types (i.e. `vin_t` and `vout_t`) and related code since crdb does not support them.
- Not erroring out when checking PostgreSQL-specific information (e.g. `synchronous_commit` and almost all the `pg_settings` dumped in the table at startup).
- Setting the `sql_sequence` SERIAL operating mode since by default `SERIAL` type columns in crdb do not create a sequence.
- Empty arrays made using the `ARRAY[]` constructor in CockroachDB require type _annotation_ (e.g. `ARRAY[]:::BIGINT[]` with ***three*** `:` characters for annotation instead of conversion), a concept that does not exist in PostgreSQL).  May need to use the literal format like `SELECT '{}'::int[];` or whatever works for both PostgreSQL and CockroachDB.
- Replaces `cardinality(arr)` with `array_length(arr,1)` since crdb does not have `cardinality`
- DELETE queries that use the `USING` clause do not work.  This affects the "rewind" queries in rewind.go.  Use a subquery.
- Deleting duplicate rows using most common methods fails to due to CRDB memory use.  A painful temporary table with a simple DISTINCT ON clause is the only way, but it is super duper slow.

Also:
- Remove the `updateAllVotes` option from `ChainDB` sync since it is always done
on the fly. For now, let `StoreBlock` keep the option to update ticket/vote spend info.
Also keep `(*ChainDB).UpdateSpendingInfoInAllTickets` as it could be a
helpful recovery and repair function.
- `HeightDB`, `HashDB`, and `HeightHashDB` now use `DBBestBlock` to get the
data from the `meta` table.  See `HeightDBLegacy`, etc. for the querying the
`blocks` table instead.
- Before sync, check the `ibd_complete` flag in the meta `table` and force indexing
and full update of `addresses` table spending info as required.
- Remove unused code and comments from dcrpg/internal code.

chappjc commented:
For the record, I went down the multi-line insert statement rabbit hole (e.g. `INSERT INTO table (col1, col2) VALUES (v11, v12), (v21, v22))`, but it did **not** improve performance over simply using a single transaction for all the inserts.  This was a rabbit hole because the go PostgreSQL drivers do not implement multi-line inserts (that I am aware of), instead advising putting many inserts into a transaction, which we are doing.

https://github.com/decred/dcrdata/compare/3625acdcd48121bfe5b80c45a7fb7710cf6bdd40...chappjc:multi-line-inserts?expand=1

dmigwi commented:
I am not sure why this is happening: 
Normal db sync from scratch isn't proceeding as expected. I will still try to get to the root cause.

```
2019-05-04 00:41:58.646 [INF] DATD: Log folder:  /home/dmigwi/.dcrdata/logs/mainnet
2019-05-04 00:41:58.646 [INF] DATD: Config file: /home/dmigwi/.dcrdata/dcrdata.conf
2019-05-04 00:41:58.646 [WRN] DATD: The --pg switch is deprecated since full-mode is now required.
2019-05-04 00:41:58.646 [INF] DATD: dcrdata version 5.0.0-pre+dev (Go version go1.12.4)
2019-05-04 00:41:58.667 [INF] RPCC: Established connection to RPC server localhost:9109
2019-05-04 00:41:58.667 [INF] DATD: Connected to dcrd (JSON-RPC API v5.0.0) on MainNet
2019-05-04 00:41:58.667 [INF] SKDB: Loading ticket pool DB. This may take a minute...
2019-05-04 00:41:58.670 [INF] SKDB: badger: All 0 tables opened in 0s
2019-05-04 00:41:58.671 [INF] SKDB: Loading all ticket pool diffs...
2019-05-04 00:41:58.671 [INF] SKDB: Successfully loaded 0 ticket pool diffs
2019-05-04 00:41:58.672 [INF] SKDB: Creating new stake DB.
2019-05-04 00:41:58.674 [INF] SKDB: Advancing ticket pool DB to tip via diffs...
2019-05-04 00:41:58.674 [INF] SKDB: Pre-populating live ticket cache and computing pool value...
2019-05-04 00:41:58.675 [WRN] SQLT: Block summary table is empty.
2019-05-04 00:41:58.675 [WRN] SQLT: Stake info table is empty.
2019-05-04 00:41:58.675 [INF] DATD: SQLite DB successfully opened: dcrdata.sqlt.db
2019-05-04 00:41:58.675 [WRN] SQLT: Block summary table is empty.
2019-05-04 00:41:58.675 [WRN] SQLT: Stake info table is empty.
2019-05-04 00:41:58.675 [INF] SQLT: SQLite block summary table height:  -1
2019-05-04 00:41:58.675 [INF] SQLT: SQLite stake info table height:     -1
2019-05-04 00:41:58.675 [INF] SQLT: StakeDatabase height:               0
2019-05-04 00:41:58.675 [INF] DATD: Setting up the Politeia's proposals clone repository. Please wait...
2019-05-04 00:42:49.086 [INF] DATD: Address cache capacity: 986895 rows, 134217728 bytes
2019-05-04 00:42:49.089 [INF] PSQL: CockroachDB CCL v19.1.0 (x86_64-unknown-linux-gnu, built 2019/04/29 18:36:40, go1.11.6)
2019-05-04 00:42:49.091 [INF] PSQL: Empty database "dcrdata". Creating tables...
2019-05-04 00:42:49.092 [INF] PSQL: Creating the "meta" table.
2019-05-04 00:42:49.099 [INF] PSQL: Creating the "blocks" table.
2019-05-04 00:42:49.115 [INF] PSQL: Creating the "vins" table.
2019-05-04 00:42:49.127 [INF] PSQL: Creating the "votes" table.
2019-05-04 00:42:49.141 [INF] PSQL: Creating the "misses" table.
2019-05-04 00:42:49.152 [INF] PSQL: Creating the "transactions" table.
2019-05-04 00:42:49.167 [INF] PSQL: Creating the "block_chain" table.
2019-05-04 00:42:49.173 [INF] PSQL: Creating the "tickets" table.
2019-05-04 00:42:49.184 [INF] PSQL: Creating the "agendas" table.
2019-05-04 00:42:49.196 [INF] PSQL: Creating the "testing" table.
2019-05-04 00:42:49.206 [INF] PSQL: Creating the "vouts" table.
2019-05-04 00:42:49.217 [INF] PSQL: Creating the "addresses" table.
2019-05-04 00:42:49.230 [INF] PSQL: Creating the "agenda_votes" table.
2019-05-04 00:42:49.241 [INF] PSQL: Creating the "proposals" table.
2019-05-04 00:42:49.257 [INF] PSQL: Creating the "proposal_votes" table.
2019-05-04 00:42:49.304 [INF] PSQL: Pre-loading unspent ticket info for InsertVote optimization.
2019-05-04 00:42:49.309 [INF] PSQL: Setting PostgreSQL DB statement timeout to 1h0m0s.
2019-05-04 00:42:49.362 [WRN] SQLT: Block summary table is empty.
2019-05-04 00:42:49.362 [INF] DATD: pgDB block summary table is empty.
2019-05-04 00:42:49.363 [WRN] SQLT: Stake info table is empty.
2019-05-04 00:42:49.395 [INF] AGDB: agendas.db version 1.0.0 was set
2019-05-04 00:42:49.403 [INF] AGDB: 4 agenda records (agendas) were updated
2019-05-04 00:42:49.409 [INF] PRDB: proposals.db version 1.0.0 was set
2019-05-04 00:42:49.439 [INF] EXPR: Mean Voting Blocks calculated: 7860
2019-05-04 00:42:49.455 [INF] EXPR: Starting WebsocketHub run loop.
2019-05-04 00:42:49.456 [INF] PUBS: Starting WebsocketHub run loop.
2019-05-04 00:42:49.529 [WRN] SQLT: Block summary table is empty.
2019-05-04 00:42:49.530 [INF] DATD: pgDB block summary table is empty.
2019-05-04 00:42:49.531 [INF] DATD: Now serving the explorer and APIs on http://127.0.0.1:7777/
2019-05-04 00:42:49.781 [INF] DATD: Starting blockchain sync...
2019-05-04 00:42:49.783 [WRN] SQLT: Block summary table is empty.
2019-05-04 00:42:49.783 [WRN] SQLT: Stake info table is empty.
2019-05-04 00:42:49.783 [INF] SQLT: Current best block (chain server):     341150
2019-05-04 00:42:49.783 [INF] SQLT: Current best block (sqlite block DB):  -1
2019-05-04 00:42:49.783 [INF] SQLT: Current best block (stakedb):          0
2019-05-04 00:42:49.784 [INF] PSQL: Tables are empty, starting fresh.
2019-05-04 00:42:49.785 [INF] PSQL: Large bulk load: Removing indexes and disabling duplicate checks.
2019-05-04 00:42:49.949 [INF] PSQL: Scanning genesis block into auxiliary chain db.
2019-05-04 00:42:49.950 [INF] SQLT: Scanning genesis block into stakedb and sqlite block db.
2019-05-04 00:42:55.694 [INF] PRDB: 30 proposal records (politeia proposals-storm) were updated
```

dmigwi commented:
I have found the bug source.
```
2019-05-04 02:24:11.409 [DBG] PSQL: SyncChainDB quit at height -1, err: StoreBlock failed: failure in InsertVoutsStmt: pq: error in argument for $9: only 1-dimension arrays supported
2019-05-04 02:24:13.023 [TRC] EXPR: Sending progress bar signal.
2019-05-04 02:24:14.629 [INF] PRDB: 0 proposal records (politeia proposals-storm) were updated
2019-05-04 02:24:15.023 [TRC] EXPR: Sending progress bar signal.
2019-05-04 02:24:17.023 [TRC] EXPR: Sending progress bar signal.
2019-05-04 02:24:19.023 [TRC] EXPR: Sending progress bar signal.
2019-05-04 02:24:21.023 [TRC] EXPR: Sending progress bar signal.
2019-05-04 02:24:23.023 [TRC] EXPR: Sending progress bar signal.
```

dmigwi commented on db/dcrpg/sync.go:
```diff
@@ -98,19 +97,19 @@ const (
+	height, err := pgb.SyncChainDB(ctx, client, updateAllAddresses, newIndexes,
+		updateExplorer, barLoad)
 	if err != nil {
 		log.Debugf("SyncChainDB quit at height %d, err: %v", height, err)
```
I think this will be more effective if it was
```Golang
   log.Errorf("SyncChainDB quit at height %d, err: %v", height, err)
```

chappjc replied:
Agreed.

Regarding the error `InsertVoutsStmt: pq: error in argument for $9: only 1-dimension arrays supported`, that's surprising since I recall resolving the exact error while preparing the PR.  What cockroach version are you using?  I'll have to check my local repo clone for unpublished commits.

dmigwi replied:
`2019-05-04 00:42:49.089 [INF] PSQL: CockroachDB CCL v19.1.0 (x86_64-unknown-linux-gnu, built 2019/04/29 18:36:40, go1.11.6)`

chappjc replied:
The version I've been using:
`CockroachDB CCL v2.1.6 (x86_64-pc-linux-gnu, built 2019/04/21 02:11:11, go1.12.4)`

How bizarre that they went from 2.1.6 to 19.0 (just released on Apr. 30).

Anyway, I just downloaded crdb v19.0 and was able to replicate the issue.

chappjc replied:
Looks like a bug in either github.com/lib/pq or CockroachDB v19.0.  It seems impossible to insert an empty `TEXT[]` with the lib/pq driver in crdb v19.0 although it worked with crdb v2.1.6.

dmigwi replied:
How about adding a recommendation that our cockroach db implementation only works with v2.1.6 to the wiki? You could also add a link on how you downloaded it.

chappjc replied:
Turns out it was a regression in CockroachDB v19: https://github.com/cockroachdb/cockroach/issues/37365#issuecomment-490238226  A fix is reportedly planned.

chappjc replied:
The fix was merged into cockroach master and is being back ported to 19.1.x: https://github.com/cockroachdb/cockroach/pull/37398
For now I've updated [the wiki](https://github.com/decred/dcrdata/wiki/CockroachDB).

chappjc commented:
@dmigwi I'm not planning to make any additional changes to this PR.

dmigwi commented:
Great! I will run tests on it today.

dmigwi commented:
I ran initial startup test from scratch on this PR and things went well but had to stop the testnet sync halfway because of how long it took. Syncing 0 - 13,000 blocks took almost 12hrs while normal pg setup full sync, takes 2 - 5hrs.
I guess the right question to ask is, when will the cockroach db implementation supersede the pg implementation?

dmigwi commented:
This comment https://github.com/decred/dcrdata/issues/1291#issuecomment-487641729 has some great ideas on how to best use cockroach db thus IMO we could maybe focus on testing the cockroach db in the ways that make it more advantageous to use than pg.

chappjc commented:
I think the answer is that is unlikely cockroach will ever give satisfactory performance.  Two options:
1. Keep this PR as a patch set in case someone wants to try it in the future.  Presently I don't think it's worth all the changes.
2. Merge the PR, and (potentially) use cockroach only for its strengths (if we need those at all) by syncing the DB with postgresql, somehow dumping to compatible sql, and importing into cockroachdb.

I'm not inclined to merge this work since it's unlikely to be used any time soon.

chappjc commented:
18 hours later, I'm leaning back towards merging it so that it may be built upon if desired.  Just requires more review and vetting with regular PostgreSQL so we don't break things.

chappjc commented:
CockroachDB 19.1.1 was released yesterday with the empty array insert fix.  https://www.cockroachlabs.com/docs/releases/v19.1.1.html

chappjc commented:
CockroachDB 19.1.1 is tested with this PR and working well.  I've updated the wiki page accordingly.  I also added a tip for disabling the WAL/raft_log write sync for performance:  https://github.com/decred/dcrdata/wiki/CockroachDB#disable-wal-sync

Benchmark for sync performance:

```
2019-05-22 08:36:29.643 [INF] PSQL: Processing blocks 1 to 500...
2019-05-22 08:36:33.308 [INF] PRDB: 31 proposal records (politeia proposals-storm) were updated
2019-05-22 08:36:49.685 [INF] PSQL: ( 21 blk/s,  218 tx/s,  218 vin/sec,  906 vout/s)
2019-05-22 08:36:55.885 [INF] PSQL: Processing blocks 501 to 1000...
2019-05-22 08:37:09.657 [INF] PSQL: ( 11 blk/s,  277 tx/s,  283 vin/sec,  871 vout/s)
2019-05-22 08:37:29.634 [INF] PSQL: ( 10 blk/s,  292 tx/s,  324 vin/sec,  914 vout/s)
2019-05-22 08:37:40.436 [INF] PSQL: Processing blocks 1001 to 1500...
2019-05-22 08:37:40.442 [INF] SQLT: Scanning blocks 1001 to 2000 (9735 live)...
2019-05-22 08:37:49.696 [INF] PSQL: ( 11 blk/s,  265 tx/s,  279 vin/sec,  907 vout/s)
2019-05-22 08:38:09.654 [INF] PSQL: ( 19 blk/s,  259 tx/s,  291 vin/sec,  808 vout/s)
2019-05-22 08:38:09.955 [INF] PSQL: Processing blocks 1501 to 2000...
2019-05-22 08:38:23.515 [INF] PSQL: Processing blocks 2001 to 2500...
2019-05-22 08:38:23.517 [INF] SQLT: Scanning blocks 2001 to 3000 (21120 live)...
2019-05-22 08:38:29.647 [INF] PSQL: ( 36 blk/s,  239 tx/s,  346 vin/sec,  706 vout/s)
2019-05-22 08:38:37.094 [INF] PSQL: Processing blocks 2501 to 3000...
2019-05-22 08:38:49.633 [INF] PSQL: ( 37 blk/s,  248 tx/s,  318 vin/sec,  707 vout/s)
2019-05-22 08:38:50.417 [INF] PSQL: Processing blocks 3001 to 3500...
2019-05-22 08:38:50.419 [INF] SQLT: Scanning blocks 3001 to 4000 (25213 live)...
<snip>
2019-05-22 08:46:57.352 [INF] PSQL: Processing blocks 11001 to 11500...
2019-05-22 08:46:57.356 [INF] SQLT: Scanning blocks 11001 to 12000 (42252 live)...
2019-05-22 08:47:09.672 [INF] PSQL: ( 14 blk/s,  215 tx/s,  401 vin/sec,  789 vout/s)
```

chappjc commented:
Really not fast, but possibly tolerable on a fast machine:

```
[INF] SQLT: resyncDBWithPoolValue completed in 9h20m33.77834783s
[INF] DATD: SQLite sync ended at height 347374
[INF] PSQL: Avg. speed: 181 tx/s, 667 vout/s
```

However, there  is an error deleting duplicates:

```
delete vins duplicate failed: failed to delete duplicate vins:  pq: root: memory budget exceeded: 10240 bytes requested, 134214656 currently allocated, 134217728 bytes in budget
```
